File: src/main/java/toker/panel/entity/Door.java
package toker.panel.entity;

import com.fasterxml.jackson.annotation.JsonView;
import toker.panel.entity.pk.DoorPK;

import javax.persistence.*;
import java.util.Collection;
import java.util.Objects;

@Entity
@Table(name = "door")
@IdClass(DoorPK.class)
public class Door {

    private Integer index;
    private Server server;
    private String name;
    private Boolean locked;
    private Collection<DoorKey> doorKeys;

    @Id
    @Column(name = "index", nullable = false)
    public Integer getIndex() {
        return index;
    }

    public void setIndex(Integer id) {
        this.index = id;
    }

    @Id
    @ManyToOne
    @JoinColumn(name = "server_id", referencedColumnName = "id", nullable = false)
    public Server getServer() {
        return server;
    }

    public void setServer(Server server) {
        this.server = server;
    }

    @Column(name = "name", nullable = false, length = 32)
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Column(name = "locked", nullable = false)
    public Boolean getLocked() {
        return locked;
    }

    public void setLocked(Boolean locked) {
        this.locked = locked;
    }

    @OneToMany(mappedBy = "door")
    @JsonView(View.DoorKeys.class)
    public Collection<DoorKey> getDoorKeys() {
        return doorKeys;
    }

    public void setDoorKeys(Collection<DoorKey> doorKeys) {
        this.doorKeys = doorKeys;
    }

    public static class View {
        public static class DoorKeys {}
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Door door = (Door) o;
        return Objects.equals(index, door.index) &&
                Objects.equals(server, door.server);
    }

    @Override
    public int hashCode() {
        return Objects.hash(index, name, locked);
    }
}

File: src/main/java/toker/panel/entity/Door.kt
package toker.panel.entity

import com.fasterxml.jackson.annotation.JsonView
import toker.panel.entity.Door.View.DoorKeys
import toker.panel.entity.pk.DoorPK
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "door")
@IdClass(DoorPK::class)
class Door(
        @Id
        @Column(name = "index", nullable = false)
        var index: Int? = null,

        @Id
        @ManyToOne
        @JoinColumn(name = "server_id", referencedColumnName = "id", nullable = false)
        var server: Server? = null,

        @Column(name = "name", nullable = false, length = 32)
        var name: String? = null,

        @Column(name = "locked", nullable = false)
        var locked: Boolean? = null
) {
    @OneToMany(mappedBy = "door")
    @JsonView(DoorKeys::class)
    var doorKeys: MutableSet<DoorKey>? = null

    interface View {
        interface DoorKeys
        interface None
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Door
        if (index != other.index) return false
        if (server != other.server) return false
        return true
    }

    override fun hashCode(): Int {
        var result = index ?: 0
        result = 31 * result + (server?.hashCode() ?: 0)
        return result
    }
}